libraries/AP_HAL/DSP.cpp

```diff
@@ -64,57 +68,40 @@ DSP::FFTWindowState::~FFTWindowState()
 {
     hal.util->free_type(_freq_bins, sizeof(float) * (_window_size), DSP_MEM_REGION);
     _freq_bins = nullptr;
+    hal.util->free_type(_derivative_freq_bins, sizeof(float) * (_bin_count), DSP_MEM_REGION);
+    _derivative_freq_bins = nullptr;
     hal.util->free_type(_hanning_window, sizeof(float) * (_window_size), DSP_MEM_REGION);
     _hanning_window = nullptr;
     hal.util->free_type(_rfft_data, sizeof(float) * (_window_size + 2), DSP_MEM_REGION);
     _rfft_data = nullptr;
 }
 
 // step 3: find the magnitudes of the complex data
-void DSP::step_cmplx_mag(FFTWindowState* fft, uint16_t start_bin, uint16_t end_bin, uint8_t harmonics, float noise_att_cutoff)
+void DSP::step_cmplx_mag(FFTWindowState* fft, uint16_t start_bin, uint16_t end_bin, float noise_att_cutoff)
 {
+    // fft->_freq_bins is populated with the complex magnitude values of the fft data
     // find the maximum power in the range we are interested in
-    float max_value = 0, max_value2 = 0, max_value3 = 0;
-    uint16_t max_bin2 = 0, max_bin3 = 0;
-    uint16_t bin_range = (end_bin - start_bin) + 1;
-    // calculate highest two peaks in the range of interest. we cannot simply find the
-    // maximum in two halves since the primary peak could extend over multiple bins
-    // instead move outwards looking for the 3dB points and then search from there
-
-    // first find the highest peak
-    vector_max_float(&fft->_freq_bins[start_bin], bin_range, &max_value, &fft->_max_energy_bin);
-    fft->_max_energy_bin += start_bin;
+    // in order to see a peak in the last bin we need to allow all the way up to the nyquist
+    const uint16_t smoothwidth = 1;
+    uint16_t bin_range = (MIN(end_bin + ((smoothwidth + 1) >> 1) + 2, fft->_bin_count) - start_bin) + 1;
+    // find the three highest peaks using a zero crossing algorithm
+    uint16_t peaks[MAX_TRACKED_PEAKS];
+    memset(peaks, 0, MAX_TRACKED_PEAKS * sizeof(uint16_t));
+    memset(fft->_peak_data, 0, sizeof(FrequencyPeakData) * MAX_TRACKED_PEAKS);
```

Comment:
can we just use sizeof(fft->_peak_data) ?

Reply:
Yes, fixed.